src/mscorlib/src/System/Text/Normalization.Unix.cs

```diff
@@ -0,0 +1,118 @@
+
+            int ret = Interop.GlobalizationInterop.IsNormalized(normalizationForm, strInput, strInput.Length);
+
+            if (ret == -1)
+            {
+                throw new ArgumentException(Environment.GetResourceString("Argument_InvalidCharSequenceNoIndex"), nameof(strInput));
+            }
+
+            return ret == 1;
+        }
+
+        public static string Normalize(this string strInput, NormalizationForm normalizationForm)
+        {
+            ValidateArguments(strInput, normalizationForm);
+
+            char[] buf = new char[strInput.Length];
+
+            for (int attempts = 2; attempts > 0; attempts--)
+            {
+                int realLen = Interop.GlobalizationInterop.NormalizeString(normalizationForm, strInput, strInput.Length, buf, buf.Length);
+
+                if (realLen == -1)
+                {
+                    throw new ArgumentException(Environment.GetResourceString("Argument_InvalidCharSequenceNoIndex"), nameof(strInput));
+                }
+
+                if (realLen <= buf.Length)
+                {
+                    return new string(buf, 0, realLen);
+                }
+
+                buf = new char[realLen];
+            }
+
+            throw new ArgumentException(Environment.GetResourceString("Argument_InvalidCharSequenceNoIndex"), nameof(strInput));
+        }
+
+        // -----------------------------
+        // ---- PAL layer ends here ----
+        // -----------------------------
+
+        private static void ValidateArguments(string strInput, NormalizationForm normalizationForm)
+        {
+            if (strInput == null)
+            {
+                throw new ArgumentNullException(nameof(strInput));
+            }
+
+            if (normalizationForm != NormalizationForm.FormC && normalizationForm != NormalizationForm.FormD &&
+                normalizationForm != NormalizationForm.FormKC && normalizationForm != NormalizationForm.FormKD)
+            {
+                throw new ArgumentException(Environment.GetResourceString("Argument_InvalidNormalizationForm"), nameof(normalizationForm));
+            }
+
+            if (HasInvalidUnicodeSequence(strInput))
+            {
+                throw new ArgumentException(Environment.GetResourceString("Argument_InvalidCharSequenceNoIndex"), nameof(strInput));
+            }
+        }
+
+        /// <summary>
+        /// ICU does not signal an error during normalization if the input string has invalid unicode,
+        /// unlike Windows (which uses the ERROR_NO_UNICODE_TRANSLATION error value to signal an error).
+        ///
+        /// We walk the string ourselves looking for these bad sequences so we can continue to throw
+        /// ArgumentException in these cases.
+        /// </summary>
+        private static bool HasInvalidUnicodeSequence(string s)
+        {
+            for (int i = 0; i < s.Length; i++)
+            {
+                char c = s[i];
```

Comment:
good idea. I'll do that.

---
In reply to: [81041918](https://github.com/dotnet/coreclr/pull/7404#discussion_r81041918) [](ancestors = 81041918)